homepage/homepage.go

```diff
@@ -1,25 +1,80 @@
 package homepage
 
 import (
+	"context"
 	"encoding/json"
 	"net/http"
+	"sync"
 
-	"github.com/ONSdigital/dp-api-clients-go/renderer"
+	"github.com/ONSdigital/dp-api-clients-go/headers"
+	"github.com/ONSdigital/dp-api-clients-go/zebedee"
 	"github.com/ONSdigital/dp-frontend-homepage-controller/mapper"
+	model "github.com/ONSdigital/dp-frontend-models/model/homepage"
 	"github.com/ONSdigital/log.go/log"
 )
 
+type MainFigure struct {
+	uri        string
+	datePeriod string
+	data       zebedee.TimeseriesMainFigure
+}
+
+//go:generate moq -out mocks_test.go -pkg homepage . ZebedeeClient RenderClient
+
+// ZebedeeClient is an interface with methods required for a zebedee client
+type ZebedeeClient interface {
+	GetTimeseriesMainFigure(ctx context.Context, userAuthToken, uri string) (m zebedee.TimeseriesMainFigure, err error)
+}
+
+// RenderClient is an interface with methods for require for rendering a template
+type RenderClient interface {
+	Do(string, []byte) ([]byte, error)
+}
+
 // Handler handles requests to homepage endpoint
-func Handler(rend renderer.Renderer) http.HandlerFunc {
+func Handler(rend RenderClient, zcli ZebedeeClient) http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
-		handle(w, req, rend)
+		handle(w, req, rend, zcli)
 	}
 }
 
-func handle(w http.ResponseWriter, req *http.Request, rend renderer.Renderer) {
+func handle(w http.ResponseWriter, req *http.Request, rend RenderClient, zcli ZebedeeClient) {
 	ctx := req.Context()
 
-	m := mapper.Homepage(ctx)
+	userAccessToken, err := headers.GetUserAuthToken(req)
+	if err == headers.ErrHeaderNotFound {
+		log.Event(ctx, "no user access token header set", log.ERROR, log.Error(err))
+		userAccessToken = ""
+	}
+	if err != nil {
+		log.Event(ctx, "error getting user access token from header", log.ERROR, log.Error(err))
+		userAccessToken = ""
+	}
+
+	mainFiguresList := getMainFiguresList()
+
+	var mappedMainFigures []model.MainFigure
+	var wg sync.WaitGroup
+	var mutex = &sync.Mutex{}
+	for _, value := range mainFiguresList {
+		wg.Add(1)
+		go func(ctx context.Context, zcli ZebedeeClient, value MainFigure) {
+			defer wg.Done()
+			zebResp, err := zcli.GetTimeseriesMainFigure(ctx, userAccessToken, value.uri)
+			if err != nil {
+				log.Event(ctx, "error getting timeseries data", log.Error(err))
+				http.Error(w, "error getting timeseries data", http.StatusBadRequest)
```

Comment:
This is not being returned so the loop of go routines will continue being iterated over. If you have 1 fail and subsequent successes this will be overwritten. This is not a pretty problem to solve.